src/controller/controller.c

```diff
@@ -444,6 +481,82 @@ static bool controller_setup_node_connection_handler(Controller *controller) {
         return true;
 }
 
+static int controller_reset_heartbeat_timer(Controller *controller, sd_event_source **event_source);
+
+static int controller_heartbeat_timer_callback(
+                sd_event_source *event_source, UNUSED uint64_t usec, void *userdata) {
+        Controller *controller = (Controller *) userdata;
+        Node *node = NULL;
+        int r = 0;
+
+        LIST_FOREACH(nodes, node, controller->nodes) {
+                uint64_t diff = 0;
+                uint64_t now = 0;
+
+                if (!node_is_online(node)) {
+                        continue;
+                }
+
+                now = get_time_micros();
+                if (now == 0) {
+                        return -EINVAL;
+                }
+
+                if (now < node->last_seen) {
+                        bc_log_error("Clock skew detected");
+                        return -EINVAL;
+                }
+
+                diff = now - node->last_seen;
+                if (diff > (uint64_t) controller->heartbeat_threshold_msec * USEC_PER_MSEC) {
+                        node_disconnect(node);
```

Comment:
It would be good to have an info log here to provide a bit of information why a node gets disconnected:
```suggestion
                       bc_log_infof("Did not receive heartbeat from node '%s' since '%d'ms. Disconnecting it...", node->name, controller->heartbeat_threshold_msec);
                        node_disconnect(node);
```

Reply:
done